tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py

```diff
@@ -0,0 +1,402 @@
+    
+    inv_t = 1/formal_integral_model.time_explosion
+
+    p = r_inner[0] + (r_outer[-1] - r_inner[0]) * p
+    idx = np.searchsorted(r_outer, p, side="right")
+
+    oz = np.zeros(size * 2)
+    oshell_id = np.zeros_like(oz, dtype=np.int64)
+
+    offset = size - idx
+
+    expected_N = (offset) * 2
+    expected_oz = np.zeros_like(oz)
+    expected_oshell_id = np.zeros_like(oshell_id).astype(np.float64)
+
+    # Calculated way to determine which shells get hit
+    expected_oshell_id[:expected_N] = (
+        np.abs(np.arange(0.5, expected_N, 1) - offset) - 0.5 + idx
+    )
+    
+    expected_oz[0:offset] = 1 + calculate_z(
+        r_outer[np.arange(size, idx, -1) - 1], p, inv_t
+    )
+    expected_oz[offset:expected_N] = 1 - calculate_z(
+        r_outer[np.arange(idx, size, 1)], p, inv_t
+    )
+
+    
+    populate_z_shells_caller[1,4](r_inner, 
+                                  r_outer, 
+                                  formal_integral_model.time_explosion, 
+                                  p, 
+                                  oz, 
+                                  oshell_id, 
+                                  actual)
+
+    #This is to find which p is being tested, so the assert statement will not crash as
+    #actual is an array and needs to be indexed
+    
+    #Need to manually do the math here to see where the error is coming from!
+    p_loc = 0
+    if p == 1e-5:
+        p_loc = 0
+    elif p == 0.5:
+        p_loc = 1
+    elif p == 0.99:
+        p_loc = 2
+    elif p == 1:
+        p_loc = 3
+
+    ntest.assert_almost_equal(actual[p_loc], expected_N)
+
+    ntest.assert_allclose(oshell_id, expected_oshell_id)
+
+    ntest.assert_allclose(oz, expected_oz, atol=1e-5)
+    
+    ntest.assert_almost_equal(oshell_id, expected_oshell_id)
+    
+    ntest.assert_almost_equal(oz, expected_oz)
+
+
+@cuda.jit
+def populate_z_shells_caller(r_inner, r_outer, time_explosion, p, oz, oshell_id, actual):
+    x = cuda.grid(1)
+    actual[x] = formal_integral_cuda.populate_z_cuda(r_inner,
+                                                     r_outer,
+                                                     time_explosion,
+                                                     p, 
+                                                     oz, 
+                                                     oshell_id)
+    
+
+@pytest.mark.parametrize(
+    "N",
+    [
+        100,
+        1000,
+        10000,
+    ],
+)
+
+def test_calculate_p_values(N):
+    r = 1.0
+    func = formal_integral_cuda.calculate_p_values
+
+    expected = r / (N - 1) * np.arange(0, N, dtype=np.float64)
+    actual = np.zeros_like(expected, dtype=np.float64)
+
+    actual[::] = func(r, N)
+    ntest.assert_allclose(actual, expected)
+    ntest.assert_almost_equal(actual, expected)
+
+
+#Make test for every function call, even the convenience ones
+
+#Review this test with Jack!
+
+@pytest.mark.parametrize(
+    ["nu", "nu_insert"],
```

Comment:
`nu` should be the full line list, you can get tyhis from one of the test numba plasmas.  You should then test on reasonable frequencies the simulation would see.